### PR TITLE
Improve SolverTuner logic and unit tests

### PR DESCRIPTION
**Pull request by AlessandroPierro**
Objective of pull request: Major update to the `SolverTuner` utility class.

## Pull request checklist

Your PR fulfills the following requirements:
- [x] Tests are part of the PR (for bug fixes / features)
- [x] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`pyb`) passes locally
- [x] Build tests (`pyb -E unit`) or (`python -m unittest`) passes locally

## Pull request type

 

 

Please check your PR type:
- [x] Feature

## New Behavior
- The new API allows the user to specify a **fitness function** that evaluates the current set of hyper-parameters, based on final cost and number of steps to solution. This is the function that is maximized by the `SolverTuner`.
- The behavior of `SolverTuner` is now fully deterministic, thanks to the `seed` argument.
- `SolverTuner` now tracks all the hyper-parameters in a structured numpy array accessible with the `SolverTuner.get_results()` method.

## Does this introduce a breaking change?

- [x] Yes

### CODE REVIEW

File: src/lava/lib/optimization/utils/solver_tuner.py [hunk 1]
```diff
@@ -26,37 +27,41 @@ def __init__(self, params_grid: ty.Dict):
     def tune(self,
              solver,
              solver_parameters: ty.Dict,
-             stopping_condition: ty.Callable[[float, int], bool] = None):
-        """Find and set solving hyperparameters for an OptimizationSolver"""
+             stopping_condition: ty.Callable[[float, int], bool] = None,
+             seed: int = 0):
+        """
+        Perform random search to optimize solver hyper-parameters.
```
Comment (phstratmann) [inline src/lava/lib/optimization/utils/solver_tuner.py:33]: For docstrings, we encourage to describe all input and output parameters in the following form:

image

File: src/lava/lib/optimization/utils/solver_tuner.py [hunk 2]
```diff
@@ -26,37 +27,41 @@ def __init__(self, params_grid: ty.Dict):
...
 
         for params in params_grid:
             hyperparameters = dict(zip(params_names, params))
             solver_parameters["hyperparameters"] = hyperparameters
-            solution = solver.solve(**solver_parameters)
-            # TODO : Implement logic for CSP problems
+            solver.solve(**solver_parameters)
             cost = solver.last_run_report["cost"]
             step_to_sol = solver.last_run_report["steps_to_solution"]
             if cost is not None and cost <= best_cost and (
```
Comment (phstratmann) [inline src/lava/lib/optimization/utils/solver_tuner.py:67]: I was thinking about this metric. Still, it feels weird to consider "best" according to our arbitrary metric (cost AND time must be minimum), instead of asking for a user input.

What's about this option:
The user defines a metric in the following way:

`def stop(cost, time) -> ty.Tuple[float, bool]:`

The first return value provides the 1D metric used to define what "best solution" means. The second return value determines if the stopping criteria has been fullfilled. For example, if I want to find the optimal solution, my function look look like

`def stop(cost, time):
    cost_opt = 10
    return cost - cost_opt, cost == cost_opt
`

Then the SovlerTuner will provide you with the hyperparameter set that minimizes the first output (cost-cost_opt), and stops if the second condition has been fulfilled.

@GaboFGuerra , what do you think?

**Review by phstratmann — CHANGES_REQUESTED** [review]
Thanks for this fix and changes!

One major point about how we may make the function more user friendly. Please wait for comment from Gabriel before implementing it. And please also let me know about your thoughts on this!

File: src/lava/lib/optimization/utils/solver_tuner.py (hunk 2 above)
Comment (AlessandroPierro) [reply src/lava/lib/optimization/utils/solver_tuner.py:67]: i agree on the "weird metric" part. I would probably go for a *fitness function* as you are suggesting, and then a target value for the fitness function to stop the execution.

```python

    def fitness(cost, steps): 
      return cost`
    
    # SolverTuner API
    def tune(self,
             solver,
             solver_parameters: ty.Dict,
             fitness: ty.Callable[[float, int], float],
             target_fitness: float = None,
             seed: int = 0):
```

File: src/lava/lib/optimization/utils/solver_tuner.py [hunk 3]
```diff
@@ -21,56 +23,89 @@ def __init__(self, params_grid: ty.Dict):
...
-                best_step_to_sol = step_to_sol
-                print(f"""Better hyperparameters configuration found!\
-                        \nHyperparameters: {best_hyperparameters}\
-                        \nBest cost: {best_cost}\
-                        \nBest step-to-solution: {best_step_to_sol}""")
-            if stopping_condition is not None and (
-                    stopping_condition(best_cost, best_step_to_sol)):
+            fitness = fitness_fn(cost, step_to_sol)
+            self._store_trial(hyperparams, cost, step_to_sol, fitness)
+            if fitness > best_fitness:
```
Comment (phstratmann) [inline src/lava/lib/optimization/utils/solver_tuner.py:83]: Minor point: I would assume that we want to minimize a function. Thus, the fitness_fn should be defined such that it is minimum for the best set of hyperparameters.

But please feel free to disagree!

File: src/lava/lib/optimization/utils/solver_tuner.py [hunk 4]
```diff
@@ -21,56 +23,89 @@ def __init__(self, params_grid: ty.Dict):
...
+                best_hyperparams = hyperparams
+                best_fitness = fitness
+                print(f"Better hyperparameters configuration found!\n"
+                      f"Hyperparameters: {best_hyperparams}")
+            if fitness_target is not None and best_fitness >= fitness_target:
                 break
-            # TODO : Add internal logging
 
-        return best_hyperparameters, best_cost, best_step_to_sol
+        success = best_fitness >= (fitness_target or -float('inf'))
```
Comment (phstratmann) [inline src/lava/lib/optimization/utils/solver_tuner.py:91]: In case that you like my comment from above, this would change to +inf

File: src/lava/lib/optimization/utils/solver_tuner.py [hunk 5]
```diff
@@ -21,56 +23,89 @@ def __init__(self, params_grid: ty.Dict):
...
+        return best_hyperparams, success
+
+    def _store_trial(self, params, cost, step_to_sol, fitness):
+        new_entry = tuple(map(lambda k: params[k], self._params_keys))
+        new_entry = new_entry + (cost, step_to_sol, fitness)
+        new_entry_array = np.array([new_entry], dtype=self._store_dtype)
+        self._store = np.concatenate([self._store, new_entry_array])
```
Comment (phstratmann) [inline src/lava/lib/optimization/utils/solver_tuner.py:98]: At the moment, you are iteratively increasing the size of the numpy array. That's resource inefficient and not what numpy arrays have been designed for.
Maybe better to initialize it to be of the numpy array to be of the same size as the number of combinations of hyperparameters. And then fill up the array.

File: src/lava/lib/optimization/utils/solver_tuner.py [hunk 6]
```diff
@@ -21,56 +23,89 @@ def __init__(self, params_grid: ty.Dict):
...
+
+    def get_results(self):
```
Comment (phstratmann) [inline src/lava/lib/optimization/utils/solver_tuner.py:100]: Slight design decision: I would prefer to have a variable self._store, and use a getter method here, maybe called results?

**Review by phstratmann — COMMENTED** [review]
Just minor comments - great changes!

File: src/lava/lib/optimization/utils/solver_tuner.py (hunk 3 above)
Comment (AlessandroPierro) [reply src/lava/lib/optimization/utils/solver_tuner.py:83]: My point in adopting the fitness (and thus maximization) nomenclature for the SolverTuner is to distinguish it from the cost minimization performed by the Solver. If we find it makes more sense to talk about minimization, though, we can change the name of the argument to something else and change the behavior accordingly. Any suggestions on this?

File: src/lava/lib/optimization/utils/solver_tuner.py (hunk 5 above)
Comment (AlessandroPierro) [reply src/lava/lib/optimization/utils/solver_tuner.py:98]: Agree that this is not the best thing performance-wise. On the other hand, we don't know how long the SolverTuner will run, so it is possible that most of the rows would remain empty. Is ok with you?

File: src/lava/lib/optimization/utils/solver_tuner.py (hunk 6 above)
Comment (AlessandroPierro) [reply src/lava/lib/optimization/utils/solver_tuner.py:100]: Sure, let me change that.

File: src/lava/lib/optimization/utils/solver_tuner.py (hunk 5 above)
Comment (phstratmann) [reply src/lava/lib/optimization/utils/solver_tuner.py:98]: That will probably be the more performant approach. Then you can still cut the numpy array once when returning the results.

**Comment by phstratmann** [comment]
One additional point that came to my attention while Chinonso and I were tuning hyperparameters:
In many cases, we don't want to search through a grid of hyperparameters. Instead, we would like to search through a set of sets of hyperparameters, i.e. we would like to try the combinations
x = 1 and y = 2
x = 5 and y = 8
but not e.g.
x=1 and y=8.

Could you add this functionality? Would be very helpful!

File: src/lava/lib/optimization/utils/solver_tuner.py [hunk 7]
```diff
@@ -1,76 +1,164 @@
...
+        self._stored_rows = 0
+        if self._store.shape[0] < len(self._search_space):
+            self._store = np.zeros(
+                len(self._search_space), dtype=self._store_dtype)
+
+        best_hyperparams = None
+        best_fitness = -float("inf")
+
+        for params in self._search_space:
+            np.random.seed(self._seed)
```
Comment (phstratmann) [inline src/lava/lib/optimization/utils/solver_tuner.py:93]: Is it useful to set the seed here again? To me it seems like the shuffling takes place in line 43, and np.random is not being called after pine 93 anymore.

File: src/lava/lib/optimization/utils/solver_tuner.py [hunk 8]
```diff
@@ -1,76 +1,164 @@
...
+        seed
+            Seed for random shuffling and numpy seeding.
         """
-        self._params_grid = params_grid
+        self._search_space = list(search_space)
+        self._params_names = params_names
+        self._shuffle = shuffle
+        self._seed = seed
+
+        if self._shuffle:
```
Comment (phstratmann) [inline src/lava/lib/optimization/utils/solver_tuner.py:42]: Minor code style: I would add the definition of the search space and it's shuffling to the tune method. No need to re-instantiate the SolverTuner each time we want a new set of Hyperparameters or change the seed.

File: src/lava/lib/optimization/utils/solver_tuner.py [hunk 9]
```diff
@@ -1,76 +1,164 @@
...
-                best_cost = cost
-                best_step_to_sol = step_to_sol
-                print(f"""Better hyperparameters configuration found!\
-                        \nHyperparameters: {best_hyperparameters}\
-                        \nBest cost: {best_cost}\
-                        \nBest step-to-solution: {best_step_to_sol}""")
-            if stopping_condition is not None and (
-                    stopping_condition(best_cost, best_step_to_sol)):
+            fitness = fitness_fn(cost, step_to_sol)
+            self._store_trial(hyperparams, cost, step_to_sol, fitness)
```
Comment (phstratmann) [inline src/lava/lib/optimization/utils/solver_tuner.py:100]: Great - love the additional features of a fitness function!

**Review by phstratmann — APPROVED** [review]
Thanks for your changes!
Two minor suggestions for changes. Then I will be pleased to see your code merged soon. Thus, let me already approve, but please still have a look at my comments before merging.

File: src/lava/lib/optimization/utils/solver_tuner.py (hunk 7 above)
Comment (AlessandroPierro) [reply src/lava/lib/optimization/utils/solver_tuner.py:93]: My point in doing this was to seed the execution of the OptimizationSolver, not the shuffling. Does it make sense to do it here? Probably the best thing to do would be to add a seed parameter to the OptimizationSolver.solve method.
